loopy/codegen/tools.py

```diff
@@ -0,0 +1,217 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from pytools import memoize_method
+from loopy.schedule import (EnterLoop, LeaveLoop, CallKernel, ReturnFromKernel,
+                            Barrier, BeginBlockItem, gather_schedule_block,
+                            ScheduleItem)
+from dataclasses import dataclass
+from typing import List, Dict
+from loopy.kernel.instruction import InstructionBase
+from loopy.kernel import LoopKernel
+from loopy.kernel.data import Iname
+
+
+__doc__ = """
+.. currentmodule:: loopy.codegen.tools
+
+.. autoclass:: ProxyKernel
+
+.. autoclass:: CodegenOperationCacheManager
+"""
+
+
+@dataclass
+class ProxyKernel:
+    """
+    Proxy to :class:`loopy.LoopKernel` to be used by
+    :class:`CodegenOperationCacheManager`.
+    """
+    instructions: List[InstructionBase]
+    schedule: List[ScheduleItem]
+    inames: Dict[str, Iname]
+
+    @property
+    @memoize_method
+    def id_to_insn(self):
+        return {insn.id: insn for insn in self.instructions}
+
+    def insn_inames(self, insn):
+        return LoopKernel.insn_inames(self, insn)
+
+    def iname_tags_of_type(self, iname, tag_type_or_types):
+        from loopy.kernel.data import filter_iname_tags_by_type
+        return filter_iname_tags_by_type(
+                self.inames[iname].tags,
+                tag_type_or_types)
+
+
+class CodegenOperationCacheManager:
+    """
+    Caches operations arising during the codegen pipeline.
+
+    .. attribute:: kernel
+
+        An instance of :class:`ProxyKernel`.
+
+    .. automethod:: with_kernel
+    .. automethod:: get_parallel_inames_in_a_callkernel
+    """
+    def __init__(self, kernel):
+        assert isinstance(kernel, ProxyKernel)
+        self.kernel = kernel
+
+    @staticmethod
+    def from_kernel(kernel):
+        assert isinstance(kernel, LoopKernel)
+        return CodegenOperationCacheManager(
+                ProxyKernel(kernel.instructions,
+                            kernel.schedule,
+                            kernel.inames))
+
+    def with_kernel(self, kernel):
+        """
+        Returns a new instance of :class:`CodegenOperationCacheManager`
+        corresponding to *kernel* if the cached variables in *self* would
+        be invalid for *kernel*, else returns *self*.
+        """
+        if ((self.kernel.instructions != kernel.instructions)
+                or (self.kernel.schedule != kernel.schedule)
+                # TODO: could be more precise by only looking the inames' attributes
+                # relevant to CodegenOperationCacheManager
+                or (self.kernel.inames != kernel.inames)):
+            # cached values are invalidated, must create a new one
```

Comment:
This should just be `KernelProxyForCodegenOperationCacheManager.__eq__`.

Reply:
Yep, better. 44735442625602d26b95fbcccc968694b8f220b3